src/StaticArraysCore.jl

```diff
@@ -44,21 +44,21 @@ const StaticVecOrMat{T} = Union{StaticVector{<:Any, T}, StaticMatrix{<:Any, <:An
 # The ::Tuple variants exist to make sure that anything that calls with a tuple
 # instead of a Tuple gets through to the constructor, so the user gets a nice
 # error message
-Base.@pure tuple_length(T::Type{<:Tuple}) = length(T.parameters)
-Base.@pure tuple_length(T::Tuple) = length(T)
-Base.@pure tuple_prod(T::Type{<:Tuple}) = length(T.parameters) == 0 ? 1 : *(T.parameters...)
-Base.@pure tuple_prod(T::Tuple) = prod(T)
-Base.@pure tuple_minimum(T::Type{<:Tuple}) = length(T.parameters) == 0 ? 0 : minimum(tuple(T.parameters...))
-Base.@pure tuple_minimum(T::Tuple) = minimum(T)
+tuple_length(T::Tuple) = length(T)
+tuple_prod(T::Tuple) = prod(T)
+tuple_minimum(T::Tuple) = minimum(T)
+
+tuple_tuple(::Type{T}) where {T<:Tuple} = (T.parameters...,)
```

Comment:
Naively it seems like `ntuple` with `Val` might be useful here, however it turns out (when checking with `Test.@inferred` and `Base.infer_effects`) that this simpler form is already OK.